packages/core/core/src/services/entity-validator/validators.ts

```diff
@@ -153,6 +154,175 @@ const addUniqueValidator = <T extends yup.AnySchema>(
+
+      // Extract the values from the repeatable data using the constructed path
+      const values = componentContext.repeatableData.map((item) => {
+        return pathToCheck.split('.').reduce((acc, key) => acc[key], item as any);
+      });
+
+      // Check if the value is repeated in the current entity
+      const isUpdatedAttributeRepeatedInThisEntity =
+        values.filter((value) => value === updatedValue).length > 1;
+
+      if (isUpdatedAttributeRepeatedInThisEntity) {
+        return false;
+      }
+    }
+
+    /**
+     * When `componentContext` is present it means we are dealing with a unique
+     * field within a component.
+     *
+     * The unique validation must consider the specific context of the
+     * component, which will always be contained within a parent content type
+     * and may also be nested within another component.
+     *
+     * We construct a query that takes into account the parent's model UID,
+     * dimensions (such as draft and publish state/locale) and excludes the current
+     * content type entity by its ID if provided.
+     */
+    const {
+      model: parentModel,
+      options: parentOptions,
+      id: excludeId,
+    } = componentContext.parentContent;
+
+    const whereConditions: Record<string, any> = {};
+    const isParentDraft = parentOptions && parentOptions.isDraft;
+
+    whereConditions.publishedAt = isParentDraft ? null : { $notNull: true };
+
+    if (parentOptions?.locale) {
+      whereConditions.locale = parentOptions.locale;
+    }
+
+    if (excludeId && !Number.isNaN(excludeId)) {
+      whereConditions.id = { $ne: excludeId };
+    }
+
+    const queryUid = parentModel.uid;
+    const queryWhere = {
+      ...componentContext.pathToComponent.reduceRight((acc, key) => ({ [key]: acc }), {
+        [updatedAttribute.name]: value,
+      }),
+
+      ...whereConditions,
+    };
+
+    // The validation should pass if there is no other record found from the query
+    return !(await strapi.db.query(queryUid).findOne({ where: queryWhere }));
+  };
+
+  const validateUniqueFieldWithinDynamicZoneComponent = async (
+    startOfPath: string
+  ): Promise<boolean> => {
+    const targetComponentUID = model.uid;
+    // Ensure that the value is unique within the dynamic zone in this entity.
+    const countOfValueInThisEntity = (componentContext?.fullDynamicZoneContent ?? []).reduce(
+      (acc, component) => {
+        if (component.__component !== targetComponentUID) {
+          return acc;
+        }
+
+        const updatedValue = component[updatedAttribute.name];
+        return updatedValue === updatedAttribute.value ? acc + 1 : acc;
+      },
+      0
+    );
+
+    if (countOfValueInThisEntity > 1) {
+      // If the value is repeated in the current entity, the validation fails.
+      return false;
+    }
+
+    // When validating a unique field within a dynamiczone, we cannot
```

Comment:
What's concretely preventing us from using the query builder directly? :thinking: 

Using populate fragments & nested filters, it _should_ be possible to reach a valid query